crates/bevy_render/src/view/window/screenshot.rs

```diff
@@ -0,0 +1,314 @@
+use std::{borrow::Cow, num::NonZeroU32, path::Path};
+
+use bevy_app::Plugin;
+use bevy_asset::{load_internal_asset, HandleUntyped};
+use bevy_ecs::prelude::*;
+use bevy_log::{error, info_span};
+use bevy_reflect::TypeUuid;
+use bevy_tasks::AsyncComputeTaskPool;
+use bevy_utils::HashMap;
+use parking_lot::Mutex;
+use thiserror::Error;
+use wgpu::{
+    CommandEncoder, Extent3d, ImageDataLayout, TextureFormat, COPY_BYTES_PER_ROW_ALIGNMENT,
+};
+
+use crate::{
+    prelude::{Image, Shader},
+    render_resource::{
+        BindGroup, BindGroupLayout, Buffer, CachedRenderPipelineId, FragmentState, PipelineCache,
+        RenderPipelineDescriptor, SpecializedRenderPipeline, SpecializedRenderPipelines, Texture,
+        VertexState,
+    },
+    renderer::RenderDevice,
+    texture::TextureFormatPixelInfo,
+    RenderApp,
+};
+
+use super::ExtractedWindows;
+
+pub type ScreenshotFn = Box<dyn FnOnce(Image) + Send + Sync>;
+
+/// A resource which allows for taking screenshots of the window.
+#[derive(Resource, Default)]
+pub struct ScreenshotManager {
+    // this is in a mutex to enable extraction with only an immutable reference
+    pub(crate) callbacks: Mutex<HashMap<Entity, ScreenshotFn>>,
+}
+
+#[derive(Error, Debug)]
+#[error("A screenshot for this window has already been requested.")]
+pub struct ScreenshotAlreadyRequestedError;
+
+impl ScreenshotManager {
+    /// Signals the renderer to take a screenshot of this frame.
+    ///
+    /// The given callback will eventually be called on one of the [`AsyncComputeTaskPool`]s threads.
+    pub fn take_screenshot(
+        &mut self,
+        window: Entity,
+        callback: impl FnOnce(Image) + Send + Sync + 'static,
+    ) -> Result<(), ScreenshotAlreadyRequestedError> {
+        self.callbacks
+            .get_mut()
+            .try_insert(window, Box::new(callback))
+            .map(|_| ())
+            .map_err(|_| ScreenshotAlreadyRequestedError)
+    }
+
+    /// Signals the renderer to take a screenshot of this frame.
+    ///
+    /// The screenshot will eventually be saved to the given path, and the format will be derived from the extension.
+    pub fn save_screenshot_to_disk(
+        &mut self,
+        window: Entity,
+        path: impl AsRef<Path>,
+    ) -> Result<(), ScreenshotAlreadyRequestedError> {
+        let path = path.as_ref().to_owned();
+        self.take_screenshot(window, move |img| match img.try_into_dynamic() {
+            Ok(dyn_img) => match image::ImageFormat::from_path(&path) {
+                Ok(format) => {
+                    // discard the alpha channel which stores brightness values when HDR is enabled to make sure
+                    // the screenshot looks right
+                    let img = dyn_img.to_rgb8();
+                    if let Err(e) = img.save_with_format(path, format) {
```

Comment:
There should probably be an info log here with the path to let people know a screenshot was taken.